connectors/kafka/src/main/java/forklift/producers/KafkaForkliftProducer.java

```diff
@@ -0,0 +1,257 @@
+    }
+
+    @Override
+    public void close() throws IOException {
+        //do nothing, the passed in KafkaProducer may be used elsewhere and should be closed by the KafkaController
+    }
+
+    private String sendForkliftWrappedMessage(String message, Map<String, String> messageProperties) throws ProducerException {
+        GenericRecord avroRecord = new GenericData.Record(forkliftSchema);
+        avroRecord.put(SCHEMA_FIELD_NAME_VALUE, message);
+        messageProperties = messageProperties == null ? new HashMap<>() : new HashMap<>(messageProperties);
+        //add the producer level properties but do not overwrite message level properties
+        for (Map.Entry<String, String> entry : this.properties.entrySet()) {
+            messageProperties.putIfAbsent(entry.getKey(), entry.getValue());
+        }
+        avroRecord.put(SCHEMA_FIELD_NAME_PROPERTIES, this.formatMap(messageProperties));
+        ProducerRecord record = new ProducerRecord<>(topic, null, avroRecord);
+        try {
+            RecordMetadata result = (RecordMetadata)kafkaProducer.send(record).get();
+            return result.topic() + "-" + result.partition() + "-" + result.offset();
+
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new ProducerException("Error sending Kafka Message", e);
+        } catch (ExecutionException e) {
+            throw new ProducerException("Error sending Kafka Message", e);
+        }
+    }
+
+    private Schema addForkliftPropertiesToSchema(Schema schema) throws IOException {
+        String originalJson = schema.toString(false);
+        JsonNode propertiesField = mapper.readTree("{\"name\":\"forkliftProperties\",\"type\":\"string\",\"default\":\"\"," +
+                                                   "\"doc\":\"Properties added to support forklift interfaces. Format is key,value entries delimited with new lines\"}");
+        ObjectNode schemaNode = (ObjectNode)mapper.readTree(originalJson);
+        ArrayNode fieldsNode = (ArrayNode)schemaNode.get("fields");
+        fieldsNode.add(propertiesField);
+        schemaNode.set("fields", fieldsNode);
+        Schema.Parser parser = new Schema.Parser();
+        return parser.parse(mapper.writeValueAsString(schemaNode));
+    }
+
+    private GenericRecord addForkliftPropertiesToAvroObject(SpecificRecord message) throws IOException {
+        //Write message to json
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        Encoder encoder = EncoderFactory.get().jsonEncoder(message.getSchema(), outputStream);
+        DatumWriter<SpecificRecord> writer = new SpecificDatumWriter<>(message.getSchema());
+        writer.write(message, encoder);
+        encoder.flush();
+        String json = new String(outputStream.toByteArray(), Charset.forName("UTF-8"));
+
+        //modify schema to include forklift properties
+        Schema modifiedSchema = avroSchemaCache.get(message.getClass());
```

Comment:
I don't like it because it exists to help optimize out the cost of doing something that doesn't seem like it should really be that expensive.

However, my judgment on that could be wrong.